linker.go

```diff
@@ -185,12 +185,53 @@ func fixupAndValidate(insns asm.Instructions) error {
 			return fmt.Errorf("instruction %d: map %s: %w", iter.Index, ins.Reference(), asm.ErrUnsatisfiedMapReference)
 		}
 
+		if err := fixupKfunc(ins); err != nil {
+			return fmt.Errorf("fixing up kfunc: %w", err) // should this return its own error type?
+		}
+
 		fixupProbeReadKernel(ins)
 	}
 
 	return nil
 }
 
+// fixupKfunc sets ins.Constant to the running kernels btf id of the btf.Func set
+// set to the instructions Metadata
+func fixupKfunc(ins *asm.Instruction) error {
+	if !ins.IsKfuncCall() {
+		return nil
+	}
+
+	// check meta, if no meta return err
+	kfm, _ := ins.Metadata.Get(kfuncMeta{}).(*btf.Func)
+	if kfm == nil {
+		return fmt.Errorf("kfunc call has no kfuncMeta")
+	}
+
+	s, err := btf.LoadKernelSpec()
+	if err != nil {
+		return fmt.Errorf("couldn't load kernel spec")
+	}
+
+	var f *btf.Func
+	if err := s.TypeByName(kfm.Name, &f); err != nil {
+		return fmt.Errorf("couldn't find kfunc in kernel spec")
+	}
+
+	// we should compare args and return type here
+	// coreAreTypesCompatible(kfm.Type, fn.Type) ?
```

Comment:
Yes, I think we'll have to export that function. How about `btf.CheckCompatibility(Type, Type) error`?

The big comment at the top of the function should move into the function body, it's too much implementation detail. We'll also have to replace `errImpossibleRelocation` with a more meaningful error, maybe even an exported one.

Reply:
How is https://github.com/cilium/ebpf/pull/896 different from exporting `coreAreTypesCompatible(Type, Type)`? cc @ti-mo

Reply:
Ah, the reason that we don't handle Typedef in `coreAreTypesCompatible` is that the function assumes that all qualifiers and Typedefs have been stripped out using `UnderlyingType`.

> Wondering if adding *Typdef to the `case *Pointer, *Array` statement would solve this?

Unfortunately not, since we want to treat the following as equal:

```go
local: Typedef{Int}
target: Int{}
```

If we just add to the switch case, we'd still hit https://github.com/cilium/ebpf/blob/5ea5368f9823c13e71146a6c2cca84ad60a5e26a/btf/core.go#L878-L880 which would fail since `Typedef != Int`.

The best fix is to create a wrapper function I think. Leave `coreAreTypesCompatible` unexported, and instead export the wrapper. The wrapper calls `Copy(local, UnderlyingType)`, etc. and then invokes `coreAreTypesCompatible`. CO-RE relocation would still call `coreAreTypesCompatible` but external callers get the more sensible behaviour.